gp/cartesian_graph.py

```diff
@@ -4,60 +4,66 @@
 
 try:
     import sympy
+    from sympy.core import expr as sympy_expr
+
+    sympy_available = True
 except ModuleNotFoundError:
-    sympy = None
+    sympy_available = False
 
 try:
     import torch  # noqa: F401
+
+    torch_available = True
 except ModuleNotFoundError:
-    torch = None
+    torch_available = False
 
 
-from .node import InputNode, OutputNode, Parameter
+from .node import Node, InputNode, OutputNode, Parameter
+from .genome import Genome
+from typing import Callable, List, Optional, Dict
 
 
 class CartesianGraph:
     """Class representing a particular Cartesian graph defined by a
     Genome.
     """
 
-    def __init__(self, genome):
+    def __init__(self, genome: Genome) -> None:
         """Init function.
 
         Parameters
         ----------
         genome: Genome
             Genome defining graph connectivity and node operations.
         """
-        self._n_outputs = None
-        self._n_inputs = None
-        self._n_columns = None
-        self._n_rows = None
-        self._nodes = None
-        self._gnome = None
```

Comment:
isn't the declaration of `_genome : Genome` missing here?

Reply:
This is the old code.

Reply:
I see, the type of `self._genome` is clear because `genome` is specified and annotated as an input variable of the `__init__`